src/traits.jl

```diff
@@ -84,7 +84,13 @@ Length(::Type{SA}) where {SA <: StaticArray} = Length(Size(SA))
 
 @pure Base.prod(::Size{S}) where {S} = prod(S)
 
-@pure @inline Base.sub2ind(::Size{S}, x::Int...) where {S} = sub2ind(S, x...)
+if !isdefined(Base, :LinearIndices) # VERSION < v"0.7-"
+    @pure @inline Base.sub2ind(::Size{S}, x::Int...) where {S} = sub2ind(S, x...)
```

Comment:
Why does this need to be `@pure`? The `LinearIndices` alternative obviously isn't, but there seems to be no code relying on this to be evaluated at inference time, or did I miss something?

Reply:
I see... mostly this lives in code generators rather than relying on code inference?

At some stage I had the intention of delegating many more things to `Size` and static indices (this would be: static ranges, static Cartesian ranges, and a static version of `LinearIndices`) and having less generated functions. However I didn't get very far. Feel free to simplify if you can prove to yourself that its not needed.

Reply:
I think we can just leave this as is for the time being. I was only concerned whether the non-`@pure` version for 0.7 might cause any troubles, but the answer seems to be no.